README.md
<!-- markdownlint-disable MD013 -->

# Cartesi Rollups Examples

This repository includes examples of decentralized applications implemented using [Cartesi Rollups](https://github.com/cartesi/rollups).

You can use online development environments such as [Gitpod](https://gitpod.io/) and [CodeSandbox](https://codesandbox.io) to open this repository directly in your browser with all [required dependencies](https://cartesi.io/docs/build-dapps/requirements) already installed. These services allow you to start experimenting immediately, but keep in mind that they are provided by third-parties and are subject to unavailability and policy changes. They may also require access to your GitHub account in order to work properly.

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/#prebuild/https://github.com/cartesi/rollups-examples/)

## Introduction

From a developer’s point of view, each decentralized application or _DApp_ is composed of two main parts: a **front-end** and a **back-end**.

The **front-end** corresponds to the user-facing interface, which for these examples will correspond to a [command-line console application](./frontend-console).

On the other hand, the **back-end** contains the business logic of the application, similar to what traditional systems would run inside a server. Its basic goal is to store and update the application state as user input is received, producing corresponding outputs. These outputs can come in the form of **vouchers** (transactions that can be carried out on layer-1, such as a transfer of assets) and **notices** (informational statements that can be verified on layer-1, such as the resulting score of a game). In addition to that, the back-end can also issue **reports**, which correspond to general information that does not need to be verifiable by third-parties, such as application logs.

When compared to traditional software development, the main difference of a Cartesi DApp is that the back-end is deployed to a decentralized network of layer-2 nodes, who continuously verify the correctness of all processing results. As a consequence, the front-end and back-end do not communicate directly with each other. Rather, the front-end sends inputs to the Cartesi Rollups framework, who in turn makes them available to the back-end instances running inside each node. After the inputs are processed by the back-end logic, the corresponding outputs are then informed back to the Rollups framework, which enforces their correctness and makes them available to the front-end and any other interested parties.

## HTTP API

As discussed above, the front-end and back-end parts of a Cartesi DApp communicate with each other through the Rollups framework.
This is accomplished in practice by using a set of HTTP interfaces, which are specified in [Cartesi's OpenAPI Interfaces repository](https://github.com/cartesi/openapi-interfaces/).

### Back-end

The DApp's back-end interacts with the Cartesi Rollups framework by retrieving processing requests and then submitting corresponding outputs.

### Front-end

The front-end part of the DApp needs to access the Cartesi Rollups framework to submit user inputs and retrieve the corresponding vouchers, notices and reports produced by the back-end.
As mentioned before, it is possible to interact with all the examples in this repository through the minimalistic and general-purpose [frontend-console](./frontend-console) application.

## Requirements

Docker version `20.10.14` with [Docker Buildkit](https://github.com/moby/buildkit) enabled is required for building the environment and executing the examples.
We recommend using [Docker Desktop](https://www.docker.com/products/docker-desktop/), which already enables Buildkit by default.
Alternatively, an environment variable with value `DOCKER_BUILDKIT=1` can also be set.

The below instructions have been tested in systems running both Linux (Ubuntu), MacOS, and Windows (using [WSL](https://docs.microsoft.com/en-us/windows/wsl/install), which is highly recommended for Windows users).

## Building

To run the examples, first clone the repository as follows:

```shell
git clone https://github.com/cartesi/rollups-examples.git
```

Then, for each example, build the required docker images:

```shell
cd <example>
docker buildx bake --load
```

This will also build the example's Cartesi Machine containing the DApp's back-end logic.

The file `<example>/dapp.json` contains some configurations for building the application. In particular, it defines the back-end's entry-point executable, along with any other files that should be made available inside the Cartesi Machine.
For certain examples, the build process also includes special [procedures for downloading and installing additional dependencies](./calculator/README.md#installing-extra-dependencies) required by the application.

## Running

Each application can be executed in Production and Host modes, as explained below.

### Production mode

In this mode, the DApp's back-end logic is executed inside a Cartesi Machine, meaning that its code is cross-compiled to the machine's RISC-V architecture. This ensures that the computation performed by the back-end is _reproducible_ and hence _verifiable_, enabling a truly trustless and decentralized execution.

After building an example as described in the previous section, you can run it in production mode by executing:

```shell
cd <example>
docker compose -f ../docker-compose.yml -f ./docker-compose.override.yml up
```

Allow some time for the infrastructure to be ready.
How much will depend on your system, but eventually the container logs will only show the continuous production of empty blocks in the local blockchain, as displayed below:

```shell
rollups-examples-hardhat-1                      | Mined empty block range #32 to #33
rollups-examples-hardhat-1                      | Mined empty block range #32 to #34
rollups-examples-hardhat-1                      | Mined empty block range #32 to #35
rollups-examples-hardhat-1                      | Mined empty block range #32 to #36
```

The environment can be shut down with the following command:

```shell
docker-compose -f ../docker-compose.yml -f ./docker-compose.override.yml down -v
```

### Host mode

The _Cartesi Rollups Host Environment_ provides the very same HTTP API as the regular one, mimicking the behavior of the actual layer-1 and layer-2 components. This way, the Cartesi Rollups infrastructure can make HTTP requests to a back-end that is running natively on localhost. This allows the developer to test and debug the back-end logic using familiar tools, such as an IDE.

The host environment can be executed with the following command:

```shell
docker compose -f ../docker-compose.yml -f ./docker-compose.override.yml -f ../docker-compose-host.yml up
```

_Note_: In production mode, rejected inputs are guaranteed to have no effect on the back-end, since in that case the Cartesi Machine is completely rolled back to its previous state. However, in host mode there is no such guarantee and it is possible for changes to persist, for instance if the DApp allows an invalid input to change a global variable or produce a database write before it is rejected.

_Note_: When running in host mode, localhost port `5004` will be used by default to allow the DApp's back-end to communicate with the Cartesi Rollups framework.

### Logging configuration

Whether in production mode or in host mode, it is possible to configure the level of logging information printed by the environment components.
The main [docker-compose.yml](./docker-compose.yml) file specifies the environment services and their configurations, and these include environment variables that can be used to control the level of logging detail for each service.

For most services, the variable `RUST_LOG` defines the log level. The possible values for it are the following: `trace`, `debug`, `info`, `warn`, and `error`.

In production mode, the `server_manager` service has two different variables to control logging levels. `SERVER_MANAGER_LOG_LEVEL` controls the level of detail for the service itself, while `REMOTE_CARTESI_MACHINE_LOG_LEVEL` controls it for the Cartesi Machine in which the back-end is executing. The possible values for these variables are slightly different: `trace`, `debug`, `info`, `warning`, `error`, and `fatal`. Note that these definitions do not affect the output printed by the back-end code itself, which has independent control of its logging level.

...

### Interactive console

It is possible to start an interactive console for the Cartesi Machine containing the application's back-end logic.
This allows you to experiment with the back-end's software stacks within its production environment, allowing you to evaluate performance and explore the most adequate technology choices for its implementation.

After the [Building](#building) step above is executed, a corresponding console Docker image is made available for that purpose. To run it and start your interactive console, type the following command:

```shell
docker run --rm -it cartesi/dapp:<example>-devel-console
```

The example's specific resources can generally be found within the `/mnt/dapp` directory.

To run the console as the `root` user, type the following command:

```shell
docker run --rm -it cartesi/dapp:<example>-devel-console run-machine-console.sh --run-as-root
```

### Advancing time

When executing an example, it is possible to advance time in order to simulate the passing of epochs. To do that, run:

```shell
curl --data '{"id":1337,"jsonrpc":"2.0","method":"evm_increaseTime","params":[864010]}' http://localhost:8545
```

## Using testnets

<!-- markdownlint-disable MD024 -->

### Interacting with deployed DApps

Several examples committed to this repository are already deployed to the Ethereum Goerli testnet: [echo-python](./echo-python/README.md), [echo-cpp](./echo-cpp/README.md), [echo-lua](./echo-lua/README.md), [echo-js](./echo-js/README.md), [echo-low-level](./echo-low-level/README.md), [sqlite](./sqlite/README.md) and [knn](./knn/README.md). In order to interact with them, you can simply use the [frontend-console](./frontend-console) tool mentioned before, but this time specifying a few connectivity configurations appropriate for the target network.

First of all, you will need to provide an account with some funds for submitting transactions. This can be accomplished by specifying a mnemonic string, and optionally an account index to use from that mnemonic.
There are a few ways to get free Goerli testnet funds using _token faucets_. Sometimes you will be required to [use social media accounts](https://goerli-faucet.mudit.blog/) to request tokens, but in other cases you can just [directly specify an account address](https://goerli-faucet.slock.it/). Do keep in mind that individual faucets are kept by third-parties, are not guaranteed to be functioning at all times, and may be discontinued.

Aside from the account to use, submitting transactions also requires you to provide the URL of an appropriate RPC gateway node for the target network. There are many options for that, and several services provide private nodes with free tiers that are more than enough for running these examples. Some options include [Alchemy](https://www.alchemy.com/), [Infura](https://infura.io/) and [Moralis](https://moralis.io/).

Finally, to query the layer-2 Cartesi Node for DApp outputs, you will need to specify the URL of its GraphQL endpoint. As a general rule, the examples deployed to Goerli have their endpoints available at `https://<example>.goerli.rollups.staging.cartesi.io/graphql`.
As such, the Echo Python DApp has its endpoint available at `https://echo-python.goerli.rollups.staging.cartesi.io/graphql`. Please refer to the [frontend-console](./frontend-console)'s documentation for details on how to use it to [send inputs](./frontend-console/README.md#sending-inputs), [list outputs](./frontend-console/README.md#listing-notices-vouchers-and-reports), [deposit tokens](./frontend-console/README.md#depositing-erc-20-tokens), and more.

### Deploying DApps

Deploying a new Cartesi DApp to a blockchain requires creating a smart contract on that network, as well as running a validator node for the DApp.

The first step is to build the DApp's back-end machine, which will produce a hash that serves as a unique identifier.

```shell
cd <example>
docker buildx bake machine --load
```

Once the machine docker image is ready, we can use it to deploy a corresponding Rollups smart contract.
This requires you to specify the account and RPC gateway to use when submitting the deploy transaction on the target network, which can be done by defining the following environment variables:

```shell
export MNEMONIC=<user sequence of twelve words>
export RPC_URL=<https://your.rpc.gateway>
```

For example, to deploy to the Goerli testnet using an Alchemy RPC node, you could execute:

```shell
export MNEMONIC=<user sequence of twelve words>
export RPC_URL=https://eth-goerli.alchemyapi.io/v2/<USER_KEY>
```

With that in place, you can submit a deploy transaction to the Cartesi DApp Factory contract on the target network by executing the following command:

```shell
DAPP_NAME=<example> docker compose --env-file ../env.<network> -f ../deploy-testnet.yml up
```

Here, `env.<network>` specifies general parameters for the target network, like its name and chain ID. In the case of Goerli, the command would be:

```shell
DAPP_NAME=<example> docker compose --env-file ../env.goerli -f ../deploy-testnet.yml up
```

This will create a file at `../deployments/<network>/<example>.json` with the deployed contract's address.
Once the command finishes, it is advisable to stop the docker compose and remove the volumes created when executing it.

```shell
DAPP_NAME=<example> docker compose --env-file ../env.<network> -f ../deploy-testnet.yml down -v
```

After that, a corresponding Cartesi Validator Node must also be instantiated in order to interact with the deployed smart contract on the target network and handle the back-end logic of the DApp.
Aside from the environment variables defined before, the node will also need a secure websocket endpoint for the RPC gateway (WSS URL).

For example, for Goerli and Alchemy, you would set the following additional variable:

```shell
export WSS_URL=wss://eth-goerli.alchemyapi.io/v2/<USER_KEY>
```

Then, the node itself can be started by running a docker compose as follows:

```shell
DAPP_NAME=<example> docker compose --env-file ../env.<network> -f ../docker-compose-testnet.yml -f ./docker-compose.override.yml up
```

Alternatively, you can also run the node on host mode by executing:

```shell
DAPP_NAME=<example> docker compose --env-file ../env.<network> -f ../docker-compose-testnet.yml -f ./docker-compose.override.yml -f ../docker-compose-host-testnet.yml up
```

## Examples

### 1. [Echo Python DApp](./echo-python)

A basic "hello world" application, this DApp's back-end is written in Python and simply copies each input received as a corresponding output notice.

### 2. [Echo C++ DApp](./echo-cpp)


CHANGELOG.md
<!-- markdownlint-disable-file MD024 -->

# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.11.1] 2022-12-07

### Changed

- Update DApp deployment information

## [0.11.0] 2022-12-06

### Added

- Support for network-specific configurations
- Support for legacy transactions (allows networks that do not support EIP-1559)
- Support for Gnosis Chiado Testnet
- Log level configurations for all Rollups services

### Changed

- Store DApp deployment information in JSON files instead of plain text
- Bump toolchain to 0.12.0
- Bump rootfs to 0.15.0
- Bump kernel to linux-5.15.63-ctsi-1
- Bump rom to 0.13.0
- Bump server-manager to 0.5.0

## [0.10.1] 2022-11-17

### Changed

- Update frontend-console's yarn.lock

## [0.10.0] 2022-11-07

### Added

- Auction DApp example
- Echo JS DApp example with easier execution in host mode and support to add NPM dependencies
- Common Contracts (usable by any example) with CartesiNFT contract

### Changed

- Bump server-manager to 0.4.2: fix issue with grpc lib
- Bump rollups to 0.7.0: increase machine deadline

## [0.9.1] 2022-10-04

### Changed

- Fix issue with Python virtual environments due to conflicts with Docker Buildx
- Bump server-manager to 0.4.1: fix issue that taints the session

## [0.9.0] 2022-09-14

### Added

- Support for deploying DApps to Arbitrum Goerli and Optimism Goerli
- Internal queue added to Inspect Server in order to serialize concurrent requests
- Quick-start recommendations using web environments Gitpod and CodeSandbox

### Changed

- Proper handling of inspect requests in SQLite DApp
- Cleaner logs by using more appropriate logging levels in offchain components
- DApp filesystem creation and DApp execution using regular 'dapp' user instead of root
- Bump toolchain to 0.11.0
- Bump rootfs to 0.14.1
- Bump kernel to linux-5.5.19-ctsi-6
- Bump rom to 0.12.0
- Bump server-manager to 0.4.0

## [0.8.1] 2022-08-24

### Changed

- Fix reference of docker image at docker-compose-testnet.yml

## [0.8.0] 2022-08-17

### Added

- Inspect Server allowing clients to directly query internal DApp state
- Echo JavaScript DApp example
- Documentation and tools to validate notices, execute vouchers, list reports and inspect DApp state

### Changed

- Major refactor in Indexer service, with significant improvements to the GraphQL API
- Major refactor in State Fold service, with significant improvements to resource consumption
- Use interval mining mode in Hardhat service for local development
- Use Ethereum Goerli instead of Polygon Mumbai for live testnet deployments
- Fix withdrawal voucher encoding in ERC-20 DApp example
- Refactor of frontend-console commands structure
- Bump server-manager to 0.3.1

## [0.7.2] 2022-07-28

### Changed

- Fix toolchain docker image reference of Lua example

## [0.7.1] 2022-07-18

### Changed

- Remove Gitpod reference while we investigate an issue using it

## [0.7.0] 2022-07-06

### Added

- Echo Low-Level C++ DApp example
- Echo Lua DApp example
- ERC-20 DApp generating vouchers for withdrawals

### Changed

- Move rollup-init call from build-machine.sh to entrypoint.sh
- Support for running nodes in host mode for DApps deployed on testnet

## [0.6.0] 2022-06-14

### Added

- Deployment mechanism for EVM-compatible testnets
- Echo Rust DApp example
- Calculator DApp example
- ERC-20 Deposit DApp example
- Basic mechanism for adding extra Python dependencies
- Usage of new exception endpoint in SQLite DApp example

### Changed

- Major reorganization of packages and build processes to reduce boilerplate code duplication
- Major refactor of the frontend-console application, originally specific for the Echo Python DApp and now generalized to support sending inputs, querying notices and depositing ERC-20 tokens for all DApps, both locally and deployed on testnets
- Deprecated usage of Hardhat tasks for interacting with applications in favor of the frontend-console

## [0.5.1] 2022-05-02

### Changed

- Fix example Dockerfiles to stop referring to local .yalc directories

## [0.5.0] 2022-04-29

### Changed

- Major update to rollups 0.2.0 with new HTTP API
- Fix m2cgen example to work with Apple Silicon

## [0.4.2] 2022-03-29

### Changed

- Fix Python DApp requirements due to changes in dependencies

## [0.4.1] 2022-03-27

### Changed

- Fix m2cgen DApp build model requirements

## [0.4.0] 2022-03-27

### Added

- Echo C++ DApp example
- Converter DApp example
- k-NN DApp example
- m2cgen DApp example

## [0.3.2] 2022-03-23

### Added

- Contributing guidelines and license

### Changed

- Updated Echo DApp polygon_mumbai deploy
- Fix template Python requirements

## [0.3.1] 2022-03-15

### Added

- Easy link in main README to open working development environment using Gitpod

### Changed

- Use Chainstack for polygon_mumbai and not Infura

## [0.3.0] 2022-03-12

### Added

- Frontend console application for Echo DApp
- SQLite DApp example
- Allow hardhat tasks to work on localhost
- Use new Rollups getNotice task to retrieve notices
- Support for submitting inputs and reading notices as UTF-8 strings

### Changed

- Locks `itsdangerous` to version 1.1.0 to avoid a crash when executing Echo DApp in host mode
- AddInput task returning submitted input's index

## [0.2.2] 2022-02-25

### Changed

- Fix make console
- Fix server-manager container version
- Bump rootfs to v0.10.0, which includes default support for Python sqlite3 core module

## [0.2.1] 2022-02-18

### Changed

- Fix error when using larger payload sizes.

## [0.2.0] 2022-01-17

### Added


integration-tests/test-util.ts
import * as protoLoader from "@grpc/proto-loader";
import * as grpc from "@grpc/grpc-js";
import { ServiceError } from "@grpc/grpc-js";
import http from "http";
import { ProtoGrpcType } from "./generated-src/proto/server-manager";
import { ServerManagerClient } from "./generated-src/proto/CartesiServerManager/ServerManager";
import { GetEpochStatusRequest } from "./generated-src/proto/CartesiServerManager/GetEpochStatusRequest";
import { GetEpochStatusResponse } from "./generated-src/proto/CartesiServerManager/GetEpochStatusResponse";
import { ChildProcess, spawn, SpawnOptions } from "child_process";
import { ProcessedInput } from "./generated-src/proto/CartesiServerManager/ProcessedInput";
import { GetSessionStatusResponse } from "./generated-src/proto/CartesiServerManager/GetSessionStatusResponse";
import { GetSessionStatusRequest } from "./generated-src/proto/CartesiServerManager/GetSessionStatusRequest";

// Utilities

const CONSOLE = `../frontend-console`;

export enum LogLevel {
    VERBOSE = "verbose",
    DEFAULT = "default",
}

export enum ErrorCodes {
    SESSION_ID_NOT_FOUND = 3,
    CONCURRENT_CALL = 10,
    NO_CONNECTION = 14,
}

export const timer = (s: number) =>
    new Promise((res) => setTimeout(res, s * 1000));

class Logger {
    enabled = true;
    logLevel = LogLevel.DEFAULT;

    log(message: any) {
        if (this.enabled) {
            console.log(message);
        }
    }

    verbose(message: any) {
        if (this.enabled && this.logLevel == LogLevel.VERBOSE) {
            console.log(message);
        }
    }
}
export const logger = new Logger();

export interface CommandOutput {
    stderr: string;
    stdout: string;
    process: ChildProcess;
}

export const spawnCommandAsync = async (
    cmd: string,
    args: string[] = [],
    options: SpawnOptions = {}
): Promise<CommandOutput> => {
    const fullCmd = `${cmd} ${args.join(" ")}`;
    logger.verbose(`${fullCmd}`);

    const process = spawn(cmd, args, options);

    process.on("error", (error) => {
        throw new Error(`Failed to execute "${fullCmd}". ${error}`);
    });

    let stderr = "";
    if (process.stderr) {
        process.stderr.on("data", (data) => {
            stderr += data;
        });
    }

    let stdout = "";
    if (process.stdout) {
        process.stdout.on("data", (data) => {
            stdout += data;
        });
    }

    let exitCode = 0;
    if (options.detached) {
        process.unref();
        process.on("close", (code: number) => {
            logger.verbose(`${fullCmd} exited with code ${code}`);
        });
    } else {
        exitCode = await new Promise((resolve) => {
            process.on("close", resolve);
        });
    }

    if (exitCode) {
        throw new Error(`${fullCmd} exited with code ${exitCode}. ${stderr}`);
    }

    return {
        stdout: stdout,
        stderr: stderr,
        process: process,
    };
};

export interface TestOptions {
    logLevel: LogLevel;
    pollingTimeout: number;
    address: string;
    environment: string;
}

export const parseArgs = (argv: string[]): TestOptions => {
    let options: TestOptions = {
        logLevel: LogLevel.DEFAULT,
        pollingTimeout: 60,
        address: "",
        environment: "prod",
    };

    let address = captureStringArg(argv, "--address");
    if (address) {
        options.address = address;
    }
    let environment = captureStringArg(argv, "--environment");
    if (environment) {
        options.environment = environment;
    }
    let timeout = captureNumberArg(argv, "--pollingTimeout");
    if (timeout >= 0) {
        options.pollingTimeout = timeout;
    }

    if (argv.includes("--verbose")) {
        options.logLevel = LogLevel.VERBOSE;
    }

    return options;
};

const captureStringArg = (argv: string[], argName: string): string => {
    let index = argv.indexOf(argName);
    if (index >= 0) {
        try {
            console.log(argv[index + 1]);
            return argv[index + 1];
        } catch (error) {
            throw new Error(`Failed to parse arguments. ${error}`);
        }
    }
    return "";
};

const captureNumberArg = (argv: string[], argName: string): number => {
    let index = argv.indexOf(argName);
    if (index >= 0) {
        try {
            let value = Number(argv[index + 1]);
            if (value >= 0) {
                return value;
            }
        } catch (error) {
            throw new Error(`Failed to parse arguments. ${error}`);
        }
    }
    return -1;
};

// General

export const sendInput = async (input: string): Promise<string> => {
    const cmd = "yarn";
    const args = ["start", "input", "send", "--payload", input];
    const options = { cwd: CONSOLE };
    const io = await spawnCommandAsync(cmd, args, options);

    return io.stdout;
};

//FIXME: keeps sending commands even after mocha dies with timeout
export const queryNotices = async (
    epoch: number,
    input: number,
    timeout: number
): Promise<any[]> => {
    const cmd = "yarn";
    const args = [
        "start",
        "notice",
        "list",
        "--epoch",
        epoch.toString(),
        "--input",
        input.toString(),
    ];
    const options = { cwd: CONSOLE };

    let count = 0;
    do {
        await timer(1);
        logger.verbose(`Attempt: ${count}`);
        const output = await spawnCommandAsync(cmd, args, options);

        if (output.stderr) {
            throw new Error(`Failed to get notices. ${output.stderr}`);
        }

        try {
            const notices = _parseNoticesOutput(output.stdout);
            if (notices.length > 0) {
                return notices;
            }
            logger.verbose(`No notice retrieved`);
            count++;
        } catch (error) {
            throw new Error(`Failed to parse notices. ${error}`);
        }
    } while (count < timeout);
    return [];
};

/**
 * Extracts Notices as a JSON array from the output generated by front-end console
 * @param output the command line output from front-end console
 * @returns JSON array with the Notices
 */
const _parseNoticesOutput = (output: string): any => {
    const splitOut = output.split("\n");
    splitOut.splice(-1); //remove the last (empty) line
    const noticesStr = splitOut.pop() ?? "[]";
    logger.verbose(`Notices: ${noticesStr}`);
    return JSON.parse(noticesStr);
};

export const isQueryServerReady = async (timeout: number): Promise<boolean> => {
    let count = 0;
    do {
        await timer(1);
        try {
            logger.verbose(`Attempt: ${count}`);
            const ping = await pingQueryServer();
            if (ping) return true;
        } catch (error: any) {
            if (error.code != "ECONNRESET" && error.code != "ECONNREFUSED") {
                throw error;
            }
        }
        count++;
    } while (count < timeout);
    return false;
};

export function pingQueryServer(): Promise<boolean> {
    const options = {
        hostname: "localhost",
        port: 4000,
        path: "/graphql",
        method: "GET",
    };

    return _sendRequest(options);
}

export const advanceEpoch = async (time = 864010): Promise<void> => {
    await _hardhatAdvanceTime(time);
    await _hardhatEvmMine();
};

const _hardhatAdvanceTime = async (time: number): Promise<void> => {
    const data = JSON.stringify({
        id: 1337,
        jsonrpc: "2.0",
        method: "evm_increaseTime",
        params: [time],
    });
    const options: http.RequestOptions = {
        hostname: "localhost",
        port: 8545,
        method: "POST",
        headers: {
            "Content-Type": "application/json",
            "Content-Length": data.length,
        },
    };

    await _sendRequest(options, data);
};

const _hardhatEvmMine = async (): Promise<void> => {
    const data = JSON.stringify({
        id: 1337,
        jsonrpc: "2.0",
        method: "evm_mine",
    });
    const options: http.RequestOptions = {
        hostname: "localhost",
        port: 8545,
        method: "POST",
        headers: {
            "Content-Type": "application/json",
            "Content-Length": data.length,
        },
    };

    await _sendRequest(options, data);
};

const _sendRequest = async (
    options: http.RequestOptions,
    data?: any
): Promise<any> => {
    return new Promise((resolve, reject) => {
        const req = http.request(options, (res: any) => {
            res.on("data", (d: any) => {
                logger.verbose(`Status code: ${res.statusCode}. Data: ${d}`);
                resolve(d);
            });
        });

        req.on("error", (error: any) => {
            logger.verbose(`Request failed: ${JSON.stringify(error)}`);
            reject(error);
        });

        if (data) {
            req.write(data);
        }
        req.end();
    });
};

export async function assertEpoch(
    epoch: number,
    client: PollingServerManagerClient,
    timeout: number
) {
    let count = 0;
    let status: GetSessionStatusResponse;
    do {
        await timer(1);
        try {
            logger.verbose(`Attempt: ${count}`);
            status = await client.getSessionStatus();
            if (status.activeEpochIndex == epoch) return true;
        } catch (error: any) {
            if (error.code != ErrorCodes.CONCURRENT_CALL) {
                throw error;
            }
        }
        count++;
    } while (count < timeout);
    return false;
}

//gRPC

export class PollingServerManagerClient {
    private client: ServerManagerClient;

    constructor(client: ServerManagerClient);
    constructor(address: string, pathToProto: string);
    constructor(...args: any[]) {
        switch (args.length) {
            case 1: {
                this.client = args[0];
                logger.verbose(`Using provided server manager gRPC client`);
                break;
            }
            case 2: {
                const address = args[0];
                const protoFile = args[1];

                const definition = protoLoader.loadSync(protoFile);
                const proto = grpc.loadPackageDefinition(
                    definition
                ) as unknown as ProtoGrpcType;

                this.client = new proto.CartesiServerManager.ServerManager(
                    address,
                    grpc.credentials.createInsecure()
                );

                logger.verbose(
                    `Created server manager gRPC client with URL: ${address}`
                );
                break;
            }
            default:
                throw new Error("Undefined constructor");
        }
    }

    async isReady(timeout: number): Promise<boolean> {
        let count = 0;
        do {
            await timer(1);
            try {
                logger.verbose(`Attempt: ${count}`);
                await this.getEpochStatus(0);
                return true;
            } catch (error: any) {
                if (
                    error.code != ErrorCodes.SESSION_ID_NOT_FOUND &&
                    error.code != ErrorCodes.CONCURRENT_CALL &&
                    error.code != ErrorCodes.NO_CONNECTION
                ) {
                    throw error;
                }
            }
            count++;
        } while (count < timeout);
        return false;
    }

    async getProcessedInputs(
        epoch: number,
        timeout: number,
        expectedInputCount: number
    ): Promise<ProcessedInput[]> {
        let count = 0;
        let status: GetEpochStatusResponse;
        do {
            await timer(1);
            try {
                logger.verbose(`Attempt: ${count}`);
                status = await this.getEpochStatus(epoch);

                let pendingInputsCount = status.pendingInputCount ?? 0;
                let processedInputsCount = status.processedInputs?.length ?? 0;

                if (
                    pendingInputsCount == 0 &&
                    processedInputsCount >= expectedInputCount
                ) {
                    return status.processedInputs ?? [];
                }
            } catch (error: any) {
                if (error.code != ErrorCodes.CONCURRENT_CALL) {
                    throw error;
                }
            }
            count++;
        } while (count < timeout);
        throw new Error("Timed out waiting Server Manager to process inputs");
    }

    async getEpochStatus(
        index: number,
        sessionId = "default_rollups_id"
    ): Promise<GetEpochStatusResponse> {
        let request: GetEpochStatusRequest = {
            sessionId: sessionId,
            epochIndex: index,
        };

        return new Promise((resolve, reject) => {
            this.client?.getEpochStatus(
                request,
                (
                    err: grpc.ServiceError | null,
                    output: GetEpochStatusResponse | undefined
                ) => {
                    if (err || !output) {
                        logger.verbose(err);
                        reject(err);
                    } else {
                        logger.verbose(output);
                        resolve(output);
                    }
                }
            );
        });
    }

    async getSessionStatus(
        sessionId = "default_rollups_id"
    ): Promise<GetSessionStatusResponse> {
        let request: GetSessionStatusRequest = {
            sessionId: sessionId,
        };

        return new Promise((resolve, reject) => {
            this.client.getSessionStatus(
                request,
                (
                    err: ServiceError | null,
                    output: GetSessionStatusResponse | undefined
                ) => {
                    if (err || !output) {
                        logger.log(err);
                        reject(err);
                    } else {
                        logger.verbose(output);
                        resolve(output);
                    }
                }
            );
        });
    }
}
